static_analyzer/engine/lsp_client.py

```
@@ -0,0 +1,612 @@
+        Returns None on timeout/empty. Handles server-initiated requests
+        and skips stray notifications automatically.
+        """
+        remaining = deadline - time.monotonic()
+        if remaining <= 0:
+            return None
+
+        try:
+            message = self._msg_queue.get(timeout=min(remaining, 1.0))
+        except queue.Empty:
+            if self._process and self._process.poll() is not None:
+                raise RuntimeError(f"LSP server process exited with code {self._process.returncode}")
```

Ruff (0.15.6)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)